src/posts/lucky_0.25.cr

```diff
@@ -0,0 +1,370 @@
+
+    ### `Avram::Operation` got a facelift
+
+    When you need to handle some logic that may not be tied directly to a single model, you can create an `Avram::Operation` to handle this. You've probably seen
+    examples in your app like the `RequestPasswordReset` and `SignInUser` that come with an Auth app.
+
+    Prior to this update, it was sort of "wild west", and completely up to you on how to implement these. For consistency sake, we recommended defining a `submit`
+    method, and then returning `yield self, value`. These operations were also limited with no way to use callbacks, file attributes, or define errors not tied to
+    an attribute.
+
+    In this release, we've created a whole new interface!
+
+    ```crystal
+    class PromoteUserValidator < Avram::Operation
+      param_key :user
+      needs user : User
+      attribute token : String
+
+      # before callbacks
+      before_run do
+        if user.has_inactive_account?
+          # custom errors
+          add_error(:user_inactive, "My custom error message")
+        end
+
+        validate_required(token)
+      end
+
+      # after callbacks
+      after_run do |user|
+        NotifyUser.new(user.email).deliver
+      end
+
+      # unified interface
+      def run
+        if user.is_promotable?
+          user
+        else
+          nil
+        end
+      end
+    end
+
+    PromoteUserValidator.run(params) do |operation, user|
+      if user
+        # the user must be promotable
+      else
+        # no user, lets check our custom error
+        operation.errors[:user_inactive]
+      end
+    end
+    ```
+
+    ### Better callbacks in `Avram::SaveOperation`
+
+    We've also made quite a few changes to the `Avram::SaveOperation` family as well.
+
+    Prior to this update, your `after_save` and `after_commit` callbacks couldn't be used with an anonymous block like how `before_save` works. Well,
+    now they can!
+
+    ```crystal
+    class SaveUser < User::SaveOperation
+      before_save do
+        # before save
+      end
+
+      after_save do |saved_user|
+        # after save
+      end
+
+      after_commit do |saved_user|
+        # after DB commit
+      end
+    end
+    ```
+
+    We have also added the ability to do these callbacks conditionally. This allows you to specify if the callbacks should actually run or not.
+
+    ```crystal
+    class SaveUser < User::SaveOperation
+      before_save :validate_card_number, if: :new_transaction?
+
+      after_save :update_billing, unless: :card_still_valid?
+
+      private def validate_card_number
+      end
+
+      private def new_transaction?
+        true
+      end
+
+      private def update_billing
+      end
+
+      private def card_still_valid?
+        false
+      end
+    end
+    ```
+
+    Lastly, there were some instances where you may be updating a record, and expecting a callback when everything was done. However, in the case of
```

Comment:
> the `after_save` and `after_commit` callbacks aren't triggered.

This has always been the case, so that's not really different. Maybe I just get rid of that whole line?